### PR TITLE
Implements followers system on the database

### PR DESCRIPTION
**Pull request by MariemBaccari**
In this PR, I implemented the followers system on the database side by:
- Adding two methods to follow and unfollow users (with tests)
- Adapting the User class: the followers and following lists are now Map<String, Boolean> attributes to maintain coherence with the Firebase Database that does not support storing Arrays properly. However, getFollowers and getFollowing still return a List of usernames so it doesn't change anything from a usability perspective since the attributes are private.

### CODE REVIEW

File: app/src/androidTest/java/com/github/sdp/mediato/DatabaseTests.java [hunk 1]
```diff
@@ -5,52 +5,143 @@
...
+                .setUsername("user_test_3")
+                .setEmail("email_test_3")
+                .setRegisterDate("19/03/2023")
+                .setLocation(new Location(3.14, 3.14))
+                .build();
     }
 
-    @After
-    //Cleans database values after each test
-    public void cleanDatabase(){
```
Comment (PH-14) [inline app/src/androidTest/java/com/github/sdp/mediato/DatabaseTests.java:52]: why was this removed ? It seems important to me that the emulator database is cleaned after tests complete

File: app/src/main/java/com/github/sdp/mediato/data/Database.java [hunk 2]
```diff
@@ -214,4 +221,45 @@ public static CompletableFuture<String> addCollection(String username, Collectio
...
+                .child(USER_PATH + myUsername + FOLLOWING_PATH + usernameToFollow).setValue(true)
+                .addOnCompleteListener(task -> {
+                    System.out.println(myUsername + " is now following " + usernameToFollow);
+                });
+        database.getReference()
+                .child(USER_PATH + usernameToFollow + FOLLOWERS_PATH + myUsername).setValue(true)
+                .addOnCompleteListener(
+                        task -> {
+                            System.out.println(usernameToFollow + " is now followed by " + myUsername);
+                        });
```
Comment (PH-14) [inline app/src/main/java/com/github/sdp/mediato/data/Database.java:242]: could those two pieces of code be used in a private method to avoid repetition ? It should be possible by giving it parameters FOLLOWING/FOLLOWER for ex.

File: app/src/main/java/com/github/sdp/mediato/data/Database.java [hunk 3]
```diff
@@ -214,4 +221,45 @@ public static CompletableFuture<String> addCollection(String username, Collectio
...
+                .child(USER_PATH + myUsername + FOLLOWING_PATH + usernameToFollow).setValue(false)
+                .addOnCompleteListener(task -> {
+                    System.out.println(myUsername + " unfollowed " + usernameToFollow);
+                });
+        database.getReference()
+                .child(USER_PATH + usernameToFollow + FOLLOWERS_PATH + myUsername).setValue(false)
+                .addOnCompleteListener(
+                        task -> {
+                            System.out.println(usernameToFollow + " was unfollowed by " + myUsername);
+                        });
```
Comment (PH-14) [inline app/src/main/java/com/github/sdp/mediato/data/Database.java:263]: Similar remark, one could event use the proposed method for all 4 pieces of code

File: app/src/main/java/com/github/sdp/mediato/data/Database.java [hunk 4]
```diff
@@ -214,4 +221,45 @@ public static CompletableFuture<String> addCollection(String username, Collectio
...
+     *
+     * @param myUsername       the current user's username
+     * @param usernameToFollow the username of the user to follow
+     */
+    public static void followUser(String myUsername, String usernameToFollow) {
+        CompletableFuture<String> future = new CompletableFuture<>();
+        database.getReference()
+                .child(USER_PATH + myUsername + FOLLOWING_PATH + usernameToFollow).setValue(true)
+                .addOnCompleteListener(task -> {
+                    System.out.println(myUsername + " is now following " + usernameToFollow);
```
Comment (PH-14) [inline app/src/main/java/com/github/sdp/mediato/data/Database.java:235]: consider converting the statement lambda to expression lambda (same remark for all  addOnCompleteListener in this PR)

File: app/src/main/java/com/github/sdp/mediato/data/Database.java [hunk 5]
```diff
@@ -214,4 +221,45 @@ public static CompletableFuture<String> addCollection(String username, Collectio
...
     }
 
+    /**
+     * Method to follow a user
+     *
+     * @param myUsername       the current user's username
+     * @param usernameToFollow the username of the user to follow
+     */
+    public static void followUser(String myUsername, String usernameToFollow) {
+        CompletableFuture<String> future = new CompletableFuture<>();
```
Comment (PH-14) [inline app/src/main/java/com/github/sdp/mediato/data/Database.java:231]: Future is never used, consider returning it and adding exception handling with it (for example when one of the user is never found).

File: app/src/main/java/com/github/sdp/mediato/data/Database.java [hunk 6]
```diff
@@ -214,4 +221,45 @@ public static CompletableFuture<String> addCollection(String username, Collectio
...
+    }
+
+    /**
+     * Method to unfollow a user
+     *
+     * @param myUsername       the current user's username
+     * @param usernameToFollow the username of the user to unfollow
+     */
+    public static void unfollowUser(String myUsername, String usernameToFollow) {
+        CompletableFuture<String> future = new CompletableFuture<>();
```
Comment (PH-14) [inline app/src/main/java/com/github/sdp/mediato/data/Database.java:252]: same remark for future expression

**Review by PH-14 — CHANGES_REQUESTED** [review]
Good implementation of the follower feature to users and database. Could use a little refactoring in the way it is handled (especially with futures), in order to make it more coherent with the rest of the database methods. Don't forget the read android studio's warnings, they can be really helpful. Apart from this, good comments and good tests as always.

File: app/src/androidTest/java/com/github/sdp/mediato/DatabaseTests.java (hunk 1 above)
Comment (MariemBaccari) [reply app/src/androidTest/java/com/github/sdp/mediato/DatabaseTests.java:52]: It was breaking some tests because they don't wait for each other to complete but now I changed it to @AfterClass that way it cleans the database after all the tests are executed instead.

File: app/src/main/java/com/github/sdp/mediato/data/Database.java (hunk 4 above)
Comment (MariemBaccari) [reply app/src/main/java/com/github/sdp/mediato/data/Database.java]: Done

File: app/src/main/java/com/github/sdp/mediato/data/Database.java (hunk 5 above)
Comment (MariemBaccari) [reply app/src/main/java/com/github/sdp/mediato/data/Database.java]: Thanks for pointing that out. I forgot to remove it (adressed in a commit).

File: app/src/main/java/com/github/sdp/mediato/data/Database.java (hunk 6 above)
Comment (MariemBaccari) [reply app/src/main/java/com/github/sdp/mediato/data/Database.java]: same answer as above !

File: app/src/main/java/com/github/sdp/mediato/data/Database.java (hunk 2 above)
Comment (MariemBaccari) [reply app/src/main/java/com/github/sdp/mediato/data/Database.java]: Thanks for the comment. I adressed this in commit [ef4f04b] by adding two helper methods, one for setting the username's value (true or false) in the follower's list and another one for the following's list. I think it's more convenient to keep the following/follower in separate methods.

**Review by PH-14 — APPROVED** [review]
